src/services/pasteEdits.ts

```diff
@@ -104,12 +107,46 @@ function pasteEdits(
                 preferences,
                 formatContext,
             };
-            forEachChild(updatedFile, function cb(node) {
-                if (isIdentifier(node) && !originalProgram?.getTypeChecker().resolveName(node.text, node, SymbolFlags.All, /*excludeGlobals*/ false)) {
-                    // generate imports
-                    importAdder.addImportForUnresolvedIdentifier(context, node, /*useAutoImportProvider*/ true);
-                }
-                node.forEachChild(cb);
+
+            // `updatedRanges` represent the new ranges that account for the offset changes caused by pasting new text and
+            // `offset` represents by how much the starting position of `pasteLocations` needs to be changed.
+            //
+            // We iterate over each updated range to get the node that wholly encloses the updated range.
+            // For each child of that node, we checked for unresolved identifiers
+            // within the updated range and try importing it.
+            let offset = 0;
+            pasteLocations.forEach((location, i) => {
+                const oldTextLength = location.end - location.pos;
+                const textToBePasted = actualPastedText ? actualPastedText[0] : pastedText[i];
+                const startPos = location.pos + offset;
+                const endPos = startPos + textToBePasted.length;
+                const range: TextRange = { pos: startPos, end: endPos };
+                offset += textToBePasted.length - oldTextLength;
+
+                const enclosingNode = findAncestor(
+                    getTokenAtPosition(context.sourceFile, range.pos),
+                    ancestorNode => rangeContainsRange(ancestorNode, range),
+                );
+                if (!enclosingNode) return;
+
+                forEachChild(enclosingNode, function importUnresolvedIdentifiers(node) {
+                    const isImportCandidate = isIdentifier(node) &&
+                        rangeContainsPosition(range, node.getStart(updatedFile)) &&
+                        !originalProgram?.getTypeChecker().resolveName(

```

Comment:
Shouldn’t this use the updated program? For example, if I have a single paste of

```ts
const b = 0;
console.log(b);
```

won’t this try to resolve `b` against the original text which didn’t have a declaration for `b`, so we’ll try to auto-import it unnecessarily?

Reply:
I didnt see this before, but originalProgram's typechecker usage is sketchy as well.. The reuse of sourceFile and parent pointer changing makes it non-viable to use.

Reply:
Can you add a test that fails when using `originalProgram`?